### PR TITLE
Virtual memory task group updates

### PR DESCRIPTION
**Pull request by daniellustig**
Working draft for the work being performed by the RISC-V Virtual Memory task group.  Not yet ready for freeze or ratification, but open for comments.

### CODE REVIEW

**Review by aswaterman — CHANGES_REQUESTED** [review]
I'm just writing this comment so there's a big red X on the merge button for the time being; take no offense :)

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1496]: I'm wondering if it would make sense to, if $pte$ is a NAPOT PTE, allow these three steps to be performed against the memory location of the first equivalent PTE.  The motivation would be a reduction in the number of stores which have to be performed.

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1718]: IMO these bits should have names so we don't have to say "bit 62" repeatedly.

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1766]: I don't see where napot_bits is defined.  If the intention is for the size to be encoded implicitly by analogy with NAPOT PMP ranges, as suggested by the 1 bit in the encoding, that requires all z to be set to 0.

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1935]: Possibly orphaned, there is no other mention of attributes in this draft?

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1011]: This is a little bit confusing descriptively, since in previous versions of the document (and possibly elsewhere in the current one, I haven't reread other chapters in the new context) the ordering was intended to subsume incoherent caching.  If the ordering is now separate from the incoherent caching that changes the interpretation of various things.  A more formal model would help.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1981]: Fixed, thanks

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex]: The virtual memory TG is considering adding other attributes, but since we haven't finalized anything, I've changed this back to Reserved

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1011]: Good feedback.  I added a new paragraph of commentary explaining why we did this: https://github.com/riscv/riscv-isa-manual/pull/611/files#diff-1673d75098c5b339d06598ff6bb126f7a429d6bed61593e9b41092ee0587c9b9R1530.  Hopefully that helps.

And yes, integration with the formal model is on the task group's to-do list, and of course if we find any discrepancies along the way, we'll make sure to update this draft before ratification.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1496]: If I understand your suggestion correctly, that would require a change to the address-translation algorithm (and the page table walker in HW) to perform a bunch of extra loads in order to find whichever NAPOT PTE is first?  We wanted to stick with the address-translation algorithm as-is, in order to make it easier for implementations to adopt this idea.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1784]: This table is the definition of napot_bits.

File: src/supervisor.tex
Comment (sorear) [reply src/supervisor.tex:1496]: No extra loads, just mask off bits corresponding to the decoded napot_bits.

(In open-source implementations and discussions) I've seen ~3 ways to handle the A/D bits:

* Punt and always raise a page fault.  No change is needed.
* Do a page table walk using normal loads, then do a CAS (or LR/SC pair) to update the final PTE was fetched.  In this case we would only need to mask off bits corresponding to napot_bits.
* Do a page table walk using loads with LR semantics, then do a SC to update the PTE.  This case does require (exactly one) additional load, but it may be undesirable to begin with since it requires taking ownership of all PTEs, even if A=D=1 already.

File: src/supervisor.tex
Comment (sorear) [reply src/supervisor.tex:1784]: OK, that makes sense now.  So we could have larger NAPOT sizes in future extensions by setting the marked bit to 0, but we couldn't have smaller NAPOT sizes without bit permutation.

Years ago there was a proposal that "small" large pages could be implemented microarchitecturally, by observing all PTEs in a cache line when a fetch from memory is done.  That works up to 32 kB for a 4 kB architectural page and 8 pointers per cache line.  So maybe we don't need smaller NAPOT pages.

File: src/supervisor.tex
Comment (sorear) [reply src/supervisor.tex:1011]: That does help, thank you.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1784]: The intention is that the bit position could in principle shift anywhere to indicate what size NAPOT region is needed, e.g.,

| N | PPN           | NAPOT region size |
|---|---------------|-------------------|
| 0 | `y yyyy yyyy` | 4KiB              |
| 1 | `y yyyy yy10` | 16KiB             |
| 1 | `y yyyy 1000` | 64KiB             |
| 1 | `y yy10 0000` | 256KiB            |
| 1 | `y 1000 0000` | 1MiB              |

We're starting with just one option, and reserving those z bits rather than forcing them to be zero now for NAPOT encoding, in case we decide later we'd rather use those z bits for something else more valuable.

Comment (sorear) [reply src/supervisor.tex:1784]: Makes sense.

File: src/supervisor.tex
Comment (sorear) [inline src/supervisor.tex:1750]: This is a backward incompatible change.  The ratified spec has:

> The V bit indicates whether the PTE is valid; if it is 0, all other bits in the PTE are don’t-cares and may be used freely by software.

On Linux you would need to create a 64 EiB swap file (may be slightly off) to create a PTE with the high bit set, but there may be other supervisors that can create such PTEs under realistic circumstances.

Comment (daniellustig) [reply src/supervisor.tex:1750]: When V=1, bit 63 is currently reserved and must be zeroed by software for forward compatibility.  The PPN field doesn't automatically expand to fill all the upper bits.  So I don't think this one is backwards-incompatible.

When V=0, yes, we would be changing the meaning of bit 63 from don't care to reserved, which is technically a backwards-incompatible change.  We don't currently have a specific plan for the V=0, C=1 combination now marked reserved, so I don't see a problem if we wanted to change that C bit back to don't care when V=0 and maintain compatibility.

Comment (sorear) [reply src/supervisor.tex:1750]: I am referring only to the V=0 C=1 combination, which is non-reserved in the ratified spec.

Comment (daniellustig) [reply src/supervisor.tex]: This table was a mistake on my part; apologies.  Hopefully the new table (with C and N) makes more sense.

**Comment by jrtc27** [comment]
CHERI-RISC-V deliberately allocated its additional PTE bits from the top down so as to try to avoid conflicting with any additional future PTE bits. Is there a reason why you chose to do the same, rather than allocating from the low bits upwards as one would normally expect?

**Comment by daniellustig** [comment]
> CHERI-RISC-V deliberately allocated its additional PTE bits from the top down so as to try to avoid conflicting with any additional future PTE bits. Is there a reason why you chose to do the same, rather than allocating from the low bits upwards as one would normally expect?

The main reason was to allow more easily for future growth of the PPN field, if necessary.  It hadn't yet come up in our group discussions that CHERI-RISC-V already used those upper bits.

**Comment by jrtc27** [comment]
> > CHERI-RISC-V deliberately allocated its additional PTE bits from the top down so as to try to avoid conflicting with any additional future PTE bits. Is there a reason why you chose to do the same, rather than allocating from the low bits upwards as one would normally expect?
> 
> The main reason was to allow more easily for future growth of the PPN field, if necessary. It hadn't yet come up in our group discussions that CHERI-RISC-V already used those upper bits.

Yeah, unfortunately we needed to do something and at the time there was nothing in the spec to accommodate non-standard permission bits (short of defining a new SATP mode entirely which is a bit overkill and there are only 2 of those free but we'd need a variant for each standard mode). If you want to see exactly what bits we're using for what it's in Section 5.3.10 Virtual Memory and Page Tables (page 165 onwards) of [CHERI ISAv8](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-951.pdf) (page 167 Figure 5.3 has a bitfield diagram to summarise).

File: src/supervisor.tex
Comment (sorear) [reply src/supervisor.tex]: New table version looks good to me.

**Comment by sorear** [comment]
Personally I think it makes the most sense for research prototypes and custom embedded systems to use the highest-order PPN bits, which are known to not be needed in the closed environment.  This avoids conflict with future standard extensions, which necessarily use bits that will not conflict with even the largest systems.  I think in many cases this is implicitly allowed by the current spec insofar as the semantics of physical addresses are completely implementation defined, but something like it could be codified.

Tangentially, I/we've been wondering how much longer risc-v is going to make it without defining a 2*XLEN-bit page table entry format (c.f. PAE, LPAE).

**Comment by jrtc27** [comment]
> Personally I think it makes the most sense for research prototypes and custom embedded systems to use the highest-order PPN bits, which are known to not be needed in the closed environment. This avoids conflict with future standard extensions, which necessarily use bits that will not conflict with even the largest systems. I think in many cases this is implicitly allowed by the current spec insofar as the semantics of physical addresses are completely implementation defined, but something like it could be codified.

Using the high PPN bits is not a good idea; it requires changing everything everywhere to mask them out in a non-standard way, and such patches must remain downstream. By using other bits we only need to mess with them in the places we really need to set/get them. For example, simple things like OpenOCD/GDB translation now needs to be patched to throw away bits that the specification says *must* be taken into account, whereas currently we can leave that code entirely untouched.

File: src/supervisor.tex
Comment (dkruckemyer-ventana) [inline src/supervisor.tex:1578]: I'm not sure this description covers the case of speculation immediately after reset, unless we're saying that the ATC entries cannot be populated until an SFENCE.VMA of the appropriate type is executed. One could certainly interpret the statement "at any time since the most recent SFENCE.VMA that subsumes that address" in line 1524 above as implying ATC entries cannot be established until the first SFENCE.VMA is executed. But I doubt that's what was intended?

File: src/supervisor.tex
Comment (dkruckemyer-ventana) [inline src/supervisor.tex:1121]: Do we state somewhere that the atomicity of an implicit PTE read (or write in the case of the A/D bits) is the same as the size of the PTE? Or are we not requiring that?

**Comment by allenjbaum** [comment]
There is no requirement that satp.MODE has any particular value at reset,
but regardless of the mode, it is treated as "bare" if you're in M-mode
(and MPRV=0, which reset requires).
So, I'm wondering if the text above should be re-worded to be

as long as *the effective virtual address mode ( indicated by* {\tt
satp}.MODE *and the hart's privilege mode)* indicates that virtual
addressing is enabled.

File: src/supervisor.tex
Comment (gfavor) [reply src/supervisor.tex:1578]: Daniel and I have a related question outstanding with Andrew whihc potentially would cover this.  Namely, can hardware start using a new satp value right away if it is written in M-mode (and start caching the results), or must hardware wait until execution has switched to the privilege mode relevant to satp (i.e S/U modes)?  (For vsatp and hgatp the same question would be wrt M/HS modes being "ok" versus VS/VU modes being required before hardware can start doing tablewalks using these.)

The latter choice clearly implies that speculative use of satp can't start immediately after reset (and not until M-mode is exited).  The former choice would also limit speculative use of satp until it at least is written by software.

At the same time, more generally, satp is undefined after reset, so it would be bad in any case to allow use of that initial satp value.

**Comment by gfavor** [comment]
To be more precise, while satp is undefined after reset, the "it" that is treated as "bare" is not satp.MODE but that the translation mode while in M-mode is always "Bare" (independent of any *atp registers).

The suggested change by Allen corresponds to one of the two options I mentioned in my last post above, i.e. hardware can't start using a new *atp value until execution enters a relevant privilege mode for that *atp register.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1591]: Resolved in latest commit: translation cannot start until the "effective privilege mode" is U or S (or hypervisor equivalent)

**Comment by daniellustig** [comment]
> Maybe delete the word "current" since it seems at odds with the preceding sentence referring to effective privilege mode (not that saying "current" is now wrong). It seems like it isn't necessary to have any qualifier to "privilege mode".

I think you're referring to the following new text?

```
when checking the U~bit, the current privilege mode is always taken to be U-mode;
```

I used "current" there to match the (pre-existing) wording within the address-translation algorithm itself:

```
  Determine if the requested memory access is
  allowed by the $pte.r$, $pte.w$, $pte.x$, and $pte.u$ bits, given the
  current privilege mode and the value of the SUM and MXR fields of
  the {\tt mstatus} register."
```

**Comment by kasanovic** [comment]
@jrtc27   We have been clear that we will not consider projects using non-standard reserved encodings as a constraint on future standards development.  That said, we should be trying to ensure there is some custom encoding available in each space to allow experimentation alongside standard encodings - though some spaces are valuable and custom is a one-way street as we promise not to reclaim anything once it's made custom.   In this case, where you want to build an orthogonal universe around the standard model and there is not enough custom space available in the space you would like (satp), I'd suggest adding additional custom CSRs holding mode bits and perhaps new instructions to allow those to be set atomically along with custom satp in your prototype.

**Comment by daniellustig** [comment]
For anyone following along: we did a bit of reorganizing.  Svnapot has been factored out into its own branch, #628.  Sv57 (#626) has been manually merged into this branch.  The plan is for the changes in this branch to be reviewed and merged as part of the Priv 1.12 review process.  Meanwhile, Svnapot will go through its own review process as a standalone extension, and that timeline won't have to block anything else in this branch.  Sorry for any logistical confusion.

File: src/supervisor.tex
Comment (gfavor) [inline src/supervisor.tex:1097]: The following non-normative comment seems inconsistent with the preceding normative statement.  The preceding requires that an SFENCE.VMA with an invalid VA has no effect (as in must have no effect).   Whereas the following comment suggests that such an SFENCE.VMA may in fact have an effect.  So which is the intention?  

The former seemingly would be preferable so that such an SFENCE.VMA could reliably be turned into an SFENCE.VMA that only has ordering effects.  But this assumes that some address (e.g. 0x7FFF_FFFF_FFFF_FFFF) will be invalid across all implementations, for now and forever.  Which seems pretty brittle (starting with wondering how J-ext pointer masking and memory tagging will interact).

So it seems like the above normative statement should be softened to indicate that "no effect" is an allowed behavior (since requiring "no effect" has no real benefit).

File: src/supervisor.tex
Comment (gfavor) [inline src/supervisor.tex:1100]: But wouldn't an implementation that ignores the value in rs1 need to over-invalidate, not under-invalidate?  Performing no invalidation would certainly surprise software that's expecting at least the specified addresse to be invalidated.

File: src/supervisor.tex
Comment (dkruckemyer-ventana) [inline src/supervisor.tex:1525]: Why does setting reserved bits in a PTE cause an access fault instead of a page fault? What will M-mode software do with this?

Seems a bit inconsistent to me since all other PTE issues end up as page faults, which can easily be delegated to S-mode.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1097]: I don't see a conflict here.  The specified behavior of SFENCE.VMA with invalid address is that it has no effect.

It is also true that SFENCE.VMA with invalid address may over-invalidate.  That's also true of any SFENCE.VMA.  It's also true in general: the implementation can invalidate address-translation cache entries any time it wants to, even when no SFENCE.VMA is executed.  That's an implementation detail at that point.  We shouldn't try to architecturally specify all the possible forms of over-invalidation.

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1525]: To paraphrase a prior discussion with @aswaterman, invalid use of defined VM features results in a page fault, while invalid physical addresses cause access faults.  Making invalid use of reserved bits trigger access faults also ensures that the behavior would be consistent if any reserved bits are later repurposed as added PPN bits.

Invalid use of NAPOT PTEs (now on #628) is a page fault because it's invalid use of a defined feature, and it is also consistent with the way misaligned superpages are handled.

In some cases different decisions could have been made, but that's the reasoning for the current choices.

Comment (dkruckemyer-ventana) [reply src/supervisor.tex:1524]: This is the first instance (I think) where the translation mechanism generates an access fault when the PTE is not well-formed (i.e. the reserved bits are non-zero).

The model seems simpler and more modular if the translation mechanism checks the PTEs for proper addresses and permissions and then hands off valid, well-formed PAs to be checked by the PMAs and PMPs. Conversely, if an access raises a page fault, software knows something was wrong with the translation process and data structures, but if an access raises an access fault, software knows that the fault lies somewhere besides the translation process and data structures.

Comment (jrtc27) [reply src/supervisor.tex:1524]: It's then inconsistent if you repurpose the bits later to be flags (e.g. as is done for NAPOT). You can't win either way, but making it be a page fault is more natural than pretending they're address bits but actually aren't, and keeps the semantics from the existing spec version, which I think should always be done unless there's a very good reason why it _can't_ be.

**Comment by jrtc27** [comment]
I hope the Zsn/Svnapot commits and reverts will be removed from the history before this lands?

**Comment by daniellustig** [comment]
> I hope the Zsn/Svnapot commits and reverts will be removed from the history before this lands?

I don't mind cleaning up the history, or rebasing, or whatever the editors would prefer, once we're ready to freeze this or actually merge it.  @aswaterman any preference?

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex:1527]: I'll forward this question to the task group for more discussion.

IIRC in earlier discussions we felt that the existing spec wasn't actually unambiguously clear about this case.  There is wording saying "Bits 63–54 are reserved for future standard use and must be zeroed by software for forward compatibility", but there isn't actually any explicit wording saying what happens if that is violated.  It doesn't show up explicitly in the translation algorithm for example.  It seemed like we needed to actually make a decision one way or the other in order to clarify a hole in the spec.  It wasn't the intention to change existing semantics.

@jrtc27 can you point to the pre-existing spec wording I'm missing that makes this clearly a page fault?  If so, then I agree we should stick with that.  If not, then we should just decide which is better and formalize it.

FWIW, spike also doesn't seem to check for this either way, while the Sail model considers this a page fault.

Comment (jrtc27) [reply src/supervisor.tex:1527]: Hm, I could have sworn it was there, but you're right, I can't find it. All the implementations I know of either ignore the bits or raise a page fault, though, and I do think a page fault would be more natural.

Comment (daniellustig) [reply src/supervisor.tex:1527]: In our discussions it was pointed out that Rocket treats this as an access fault rather than a page fault: https://github.com/chipsalliance/rocket-chip/blob/8f9149faf01fec9725d08c31c9c0bb5341f5e999/src/main/scala/rocket/PTW.scala#L154

So unfortunately we seem to have both behaviors in the wild already.

**Comment by aswaterman** [comment]
> > I hope the Zsn/Svnapot commits and reverts will be removed from the history before this lands?
> 
> I don't mind cleaning up the history, or rebasing, or whatever the editors would prefer, once we're ready to freeze this or actually merge it. @aswaterman any preference?

I'd just squash & merge the PR.  (This obviously destroys more history than requested, but anyone interested in the history can come to this page to see it.)

File: src/supervisor.tex
Comment (daniellustig) [reply src/supervisor.tex]: After discussing, we have indeed changed this to a page fault.  The expectation is that all misconfigurations of a page table entry (use of reserved bits, misaligned superpages, invalid use of r/w/x, invalid use of Svnapot) will result in page faults, including if/when we start using more reserved bits for other purposes.  Access faults will remain for things like unsupported misaligned accesses or invalid physical addresses, which aren't specific to the translation process.

Note that this rule will apply once priv 1.12 is ratified.  It is not a retroactive change, as implementations raising access faults exist in the wild.

Thanks all for the feedback!

Comment (ingallsj) [reply src/supervisor.tex]: tracking Rocket PTW change request: https://github.com/chipsalliance/rocket-chip/issues/2795

Comment (mickflemm) [reply src/supervisor.tex]: If an implementation today raises a fault when reserved PTE bits are used, then it'll be incompatible with future software making use of those bits. Hw should just ignore reserved bits, or else we 'll need to fill the sw with checks on the critical path so that we choose PTE formats based on priv. spec revisions, or chose PTE formats at compile-time (meaning we can't have one kernel image for all systems).

It also seems a bit inconsistent to me, for CSRs for example WPRI fields ignore writes to reserved bits (hardwired to 0), WARL fields also ignore them ("Implementations will not raise an exception") and WLRL fields "are permitted but not required to raise an exception". If we write an invalid/reserved SATP value, it's ignored, if we use an invalid/reserved bit on sfence.vma 's rs2 it's ignored, and TBH I couldn't find a case on the current Priv. spec where writing a reserved bit/field requires the hw to raise an exception. Another inconsistency is that there are not just reserved PTE bits but also reserved combinations of standard PTEs, e.g. RW=01, or reserved combinations of D,A,U bits on non-leaf PTEs. If we are about to return a fault for invalid PTEs, we should include those cases as well.

Comment (daniellustig) [reply src/supervisor.tex]: This was already debated in the discussion above and in the virtual memory TG.  As of Priv 1.12, once this change is ratified, any invalid encoding (use of upper reserved bits, RW=01, reserved combinations of D,A,U) will trigger a page fault.  Plus, as discussed above, most implementations already raise a page fault or an access fault, and we'd like to make this consistent (as a page fault) moving forward.  Is there new information here that should cause us to open this up again?

What new checks would have to be performed on the critical path?  The OS obviously shouldn't attempt to use the new features if they aren't supported by a given implementation, and in both extensions we've proposed so far the zero(s) encoding is backwards compatible with the existing behavior.

Comment (mickflemm) [reply src/supervisor.tex]: I get it that invalid use of _defined_ VM features should result a fault, but we are talking about features to be defined at a later spec revision, using reserved bits currently unused. Why should we raise an exception on those where we can just ignore them ? And what is sw supposed to do in this case ? We won't even be able to call the trap handler to print the crash info, we'll just hang. We are turning this case to a fatal error for no apparent reason (the hw can work fine by ignoring those bits). As for the critical path and the changes needed, on Linux the core page-table related functions are defined on header files (in our case arch/riscv/include/asm/pgtable.h) so that they can be inlined, and the idea is to keep them as simple as possible (we do mappings all the time, this needs to be as efficient as possible), if we start adding checks there for every new PTE bit we decide to add in the future e.g. on set_pte_at we'll end up having something like powerpc's set_pte_filter (https://elixir.bootlin.com/linux/latest/source/arch/powerpc/mm/pgtable.c#L103) only worse (and that's the only arch I could find that follows this approach, others just use a different header file at compile time, something we want to avoid since we want to have one image that runs everywhere). I'm not saying it's not possible or wrong, I'm saying it's a bit ugly and sub-optimal sw-wise.

As for the consistency, the text still says "For non-leaf PTEs, the D, A, and U bits are reserved for future standard use and must be cleared by software for forward compatibility.", also "Writable pages must also be marked readable; the contrary combinations are reserved for future use", shouldn't we also say that a page fault must occur in those cases like we do for reserved PTE bits on Sv39 ? It's only mentioned on the translation process algorithm.

**Comment by allenjbaum** [comment]
Why shouldn't using features on HW that doesn't implement them be
considered a buggy OS, and get an exception.
Why shouldn't that be considered a corrupt PTE?
A custom extension could supersede that, I suppose.
But - why couldn't the trap handler panic, print an error message, and shut
down the process?
Why will the error message (to console output, I assume) hang? (if it's a
user page table, and not an OS page table - but even kernel panics can
print an error message)
I'm also missing why you expect a single header file, unconditioned by the
architecture type.
I'm not saying you are wrong here - but your arguments need a lot more
backup.

File: src/priv-preface.tex
Comment (pdonahue-ventana) [inline src/priv-preface.tex:48]: should or must?

**Comment by daniellustig** [comment]
Content merged via #759.  Closing.

**Pull request closed without merging**
